File: src/java/magma/option/Some.java
package magma.option;

import magma.Tuple;

import java.util.function.Consumer;
import java.util.function.Function;
import java.util.function.Supplier;

public record Some<T>(T value) implements Option<T> {
    @Override
    public <R> R match(Function<T, R> ifPresent, Supplier<R> ifEmpty) {
        return ifPresent.apply(this.value);
    }

    @Override
    public void ifPresent(Consumer<T> consumer) {
        consumer.accept(this.value);
    }

    @Override
    public <R> Option<R> map(Function<T, R> mapper) {
        return new Some<>(mapper.apply(this.value));
    }

    @Override
    public T orElseGet(Supplier<T> other) {
        return this.value;
    }

    @Override
    public Tuple<Boolean, T> toTuple(T other) {
        return new Tuple<>(true, this.value);
    }

    @Override
    public Option<T> or(Supplier<Option<T>> other) {
        return this;
    }

    @Override
    public boolean isEmpty() {
        return false;
    }

    @Override
    public boolean isPresent() {
        return true;
    }

    @Override
    public T unwrap() {
        return this.value;
    }

    @Override
    public T orElse(T other) {
        return this.value;
    }

    @Override
    public <R> Option<R> flatMap(Function<T, Option<R>> mapper) {
        return mapper.apply(this.value);
    }
}

File: src/c/magma/option/Some.c
import magma.Tuple;
import java.util.function.Consumer;
import java.util.function.Function;
import java.util.function.Predicate;
import java.util.function.Supplier;
struct Some<T>(T value) implements Option<T> {
	R match(Function<T, R> ifPresent, Supplier<R> ifEmpty){
		return ifPresent.apply(this.value);
	}
	void ifPresent(Consumer<T> consumer){
		consumer.accept(this.value);
	}
	Option<R> map(Function<T, R> mapper){
		return Some<>(mapper.apply(this.value));
	}
	T orElseGet(Supplier<T> other){
		return this.value;
	}
	T> toTuple(T other){
		return Tuple<>(true, this.value);
	}
	Option<T> or(Supplier<Option<T>> other){
		return this;
	}
	boolean isEmpty(){
		return false;
	}
	boolean isPresent(){
		return true;
	}
	T unwrap(){
		return this.value;
	}
	T orElse(T other){
		return this.value;
	}
	Option<R> flatMap(Function<T, Option<R>> mapper){
		return mapper.apply(this.value);
	}
	Option<T> filter(Predicate<T> predicate){
		return predicate.test(this.value) ? this : new None<>();
	}
}